websocket.go

```diff
@@ -0,0 +1,277 @@
+			log.Printf("Unknown operation: %s", msg["op"])
+		}
+
+		// Reset the timeout for the next read
+		cancel() // Cancel the previous context
+		ctx, cancel = context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel() // Ensure the new context is canceled on all paths
+	}
+}
+
+func (p *WebSocketProxy) handleSubscribe(conn *websocket.Conn, msg map[string]any) {
+	swapIDs, ok := msg["args"].([]any)
+	if !ok {
+		log.Printf("Invalid subscribe message: %v", msg)
+		return
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for _, swapID := range swapIDs {
+		id, ok := swapID.(string)
+		if !ok {
+			log.Printf("Invalid swap ID: %v", swapID)
+			continue
+		}
+
+		// Add client to swap ID's subscribers
+		if p.subscribers[id] == nil {
+			p.subscribers[id] = make(map[*websocket.Conn]bool)
+		}
+		p.subscribers[id][conn] = true
+
+		// Add swap ID to client's subscriptions
+		p.clients[conn][id] = true
+	}
+
+	// Forward subscribe message to upstream
+	if p.upstream != nil {
+		// Marshal the message into JSON ([]byte)
+		messageBytes, err := json.Marshal(msg)
+		if err != nil {
+			log.Printf("Failed to marshal subscribe message: %v", err)
+			return
+		}
+
+		// Write the message to the upstream WebSocket
+		err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+		if err != nil {
+			log.Printf("Failed to forward subscribe message to upstream: %v", err)
+		}
+	}
+}
+
+func (p *WebSocketProxy) handleUnsubscribe(conn *websocket.Conn, msg map[string]any) {
+	swapIDs, ok := msg["args"].([]any)
+	if !ok {
+		log.Printf("Invalid unsubscribe message: %v", msg)
+		return
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for _, swapID := range swapIDs {
+		id, ok := swapID.(string)
+		if !ok {
+			log.Printf("Invalid swap ID: %v", swapID)
+			continue
+		}
+
+		// Remove client from swap ID's subscribers
+		delete(p.subscribers[id], conn)
+		if len(p.subscribers[id]) == 0 {
+			delete(p.subscribers, id)
+		}
+
+		// Remove swap ID from client's subscriptions
+		delete(p.clients[conn], id)
+	}
+
+	// Forward unsubscribe message to upstream
+	if p.upstream != nil {
+		// Marshal the message into JSON ([]byte)
+		messageBytes, err := json.Marshal(msg)
+		if err != nil {
+			log.Printf("Failed to marshal unsubscribe message: %v", err)
+			return
+		}
+
+		// Write the message to the upstream WebSocket
+		err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+		if err != nil {
+			log.Printf("Failed to forward unsubscribe message to upstream: %v", err)
+		}
+	}
+}
+
+func (p *WebSocketProxy) Start() {
+	// Connect to upstream WebSocket server
+	err := p.connectToUpstream()
+	if err != nil {
+		log.Fatalf("Failed to connect to upstream WebSocket: %v", err)
+	}
+
+	// Start a goroutine to listen for messages from the upstream server
+	go func() {
+		for {
+			_, message, err := p.upstream.Read(context.Background())
+			if err != nil {
+				log.Printf("Upstream WebSocket read error: %v", err)
+				break
```

Comment:
This silently breaks the for loop, without recovery. Should we try to reconnect? That will complicate the other references to 'upstream' a bit.